R/DBI-connection-transaction.R

```diff
@@ -50,6 +51,103 @@ setMethod("dbCommit", "Pool", function(conn, ...) {
+#' in its body, a database connection. So, you can use connection methods
+#' wihtout ever having to check out a connection. But you can also use
+#' arbitrary R code inside the \code{func}'s body. This function will be
+#' called once we fetch a connection from the pool. Once the function
+#' returns, we release the connection back to the pool.
+#'
+#' Like its DBI sister \code{\link[DBI]{dbWithTransaction}}, this function
+#' calls \code{dbBegin()} before executing the code, and \code{dbCommit()}
+#' after successful completion, or \code{dbRollback()} in case of an error.
+#' This means that calling \code{poolWithTransaction} always has side
+#' effects, namely to commit or roll back the code executed when \code{func}
+#' is called. In addition, if you modify the local R environment from within
+#' \code{func} (e.g. setting global variables, writing to disk), these
+#' chnages will persist after the function has returned.
+#'
+#' Also, like \code{\link[DBI]{dbWithTransaction}}, there is also a special
+#' function called \code{dbBreak()} that allows for an early, silent exit
+#' with rollback. It can be called only from inside \code{poolWithTransaction}.
+#'
+#' @param pool The pool object to fetch the connection from.
+#' @param func A function that has one argument, \code{conn} (a database
+#'   connection checked out from \code{pool}).
+#'
+#' @return \code{func}'s return value.
+#'
+#' @examples
+#' pool <- dbPool(RSQLite::SQLite(), ":memory:")
+#'
+#' dbWriteTable(pool, "cars", head(cars, 3))
+#' dbReadTable(pool, "cars")   # there are 3 rows
+#'
+#' ## successful transaction
+#' poolWithTransaction(pool, function(conn) {
+#'   dbExecute(conn, "INSERT INTO cars (speed, dist) VALUES (1, 1);")
+#'   dbExecute(conn, "INSERT INTO cars (speed, dist) VALUES (2, 2);")
+#'   dbExecute(conn, "INSERT INTO cars (speed, dist) VALUES (3, 3);")
+#' })
+#' dbReadTable(pool, "cars")   # there are now 6 rows
+#'
+#' ## failed transaction -- note the missing comma
+#' tryCatch(
+#'   poolWithTransaction(pool, function(conn) {
+#'     dbExecute(conn, "INSERT INTO cars (speed, dist) VALUES (1, 1);")
+#'     dbExecute(conn, "INSERT INTO cars (speed dist) VALUES (2, 2);")
+#'     dbExecute(conn, "INSERT INTO cars (speed, dist) VALUES (3, 3);")
+#'   }),
+#'   error = identity
+#' )
+#' dbReadTable(pool, "cars")   # still 6 rows
+#'
+#' ## early exit, silently
+#' poolWithTransaction(pool, function(conn) {
+#'   dbExecute(conn, "INSERT INTO cars (speed, dist) VALUES (1, 1);")
+#'   dbExecute(conn, "INSERT INTO cars (speed, dist) VALUES (2, 2);")
+#'   if (nrow(dbReadTable(conn, "cars")) > 7) dbBreak()
+#'   dbExecute(conn, "INSERT INTO cars (speed, dist) VALUES (3, 3);")
+#' })
+#' dbReadTable(pool, "cars")   # still 6 rows
+#'
+#' @export
+poolWithTransaction <- function(pool, func) {
+  conn <- poolCheckout(pool)
+  on.exit(poolReturn(conn))
+  DBI::dbWithTransaction(conn, func(conn))
+}
+
+
+#' @export
+#' @rdname poolWithTransaction
+dbBreak <- DBI::dbBreak
```

Comment:
Well, most code that is used is `DBI::dbWithTransaction` can be copied as is into the body of `func` (the second argument in `pool::poolWithTransaction`). And, I figured, that most people would have it originally just as `dbBreak()` (instead of `DBI::dbBreak()`). So, in order to make the transition from one function to the other as smooth as possible (and lessen frustrating error messages), I thought in just setting `pool::dbBreak` to `DBI::dbBreak`. What do you think, not worth it?